### PR TITLE
[shardformer] support interleaved pipeline

### LINKED ISSUE #4447: [pipeline] refactor interleaved schedule
The previous version of interleaved schedule depends on gpc, which would be removed soon.
And there should be a new implementation of interleaved schedule.

### PR DESCRIPTION
**Pull request by Gy-Lu**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

closed #4447 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

1. This PR mainly supports interleaved schedule with the new coordinate.
2. Weaken the p2p ops. Things like examing if the first/last stage should be done in schedule.
3. Support a ring group, e.g. the prev_rank of rank0 is the last rank, and the next_rank of the last rank is rank0.
4. Remove useless codes.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/stage_manager.py [hunk 1]
```diff
@@ -17,28 +17,24 @@ class PipelineStageManager:
...
         coord = self.pg_mesh.coordinate()
-        if self.stage > 0:
-            prev_coord = coord[: self.pipeline_axis] + \
-                (coord[self.pipeline_axis] - 1,) + coord[self.pipeline_axis + 1:]
-            self.prev_rank = self.pg_mesh.ravel(prev_coord, self.pg_mesh.shape)
-        if self.stage < self.num_stages - 1:
-            next_coord = coord[: self.pipeline_axis] + \
-                (coord[self.pipeline_axis] + 1,) + coord[self.pipeline_axis + 1:]
-            self.next_rank = self.pg_mesh.ravel(next_coord, self.pg_mesh.shape)
+        # the prev rank of rank0 is the last rank
```
Comment (ver217) [inline colossalai/pipeline/stage_manager.py:30]: What's the benifit of this design?

Comment (Gy-Lu) [reply colossalai/pipeline/stage_manager.py:30]: For interleaved, the fwd output of the last rank(not the last stage) can be sent to rank0 directly.
The bwd grad of rank0(not the first stage) can be sent to the last rank.

File: colossalai/pipeline/schedule/interleaved_pp.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        if device is not None:
+            batch = tree_map(partial(to_device, device=device), batch)
+        self.batch = batch
+        self.batch_size = get_batch_size(batch)
+        self.microbatch_offset = [0 for _ in range(self.num_model_chunks)]
+        assert self.batch_size % self.num_microbatches == 0, \
+            "Batch size should divided by the number of microbatches"
+        self.microbatch_size = self.batch_size // self.num_microbatches
+
+    def load_micro_batch(self, model_chunk_id) -> Any:
```
Comment (ver217) [inline colossalai/pipeline/schedule/interleaved_pp.py:49]: Please add type hint for `model_chunk_id`

File: colossalai/pipeline/schedule/interleaved_pp.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+        """
+
+        # Retain the grad on the input_obj.
+        tree_map(retain_grad, input_obj)
+
+        # Backward pass.
+        if output_obj_grad is None:
+            optimizer.backward(output_obj)
+        else:
+            for k, grad in output_obj_grad.items():
```
Comment (ver217) [inline colossalai/pipeline/schedule/interleaved_pp.py:216]: This is updated to handle output which does not need backward, see https://github.com/hpcaitech/ColossalAI/blob/d20dceb9a3d1bdcb2376201220f49fca7c7c1be9/colossalai/pipeline/schedule/one_f_one_b.py#L110

File: colossalai/pipeline/schedule/interleaved_pp.py (hunk 2 above)
Comment (Gy-Lu) [reply colossalai/pipeline/schedule/interleaved_pp.py]: Done

File: colossalai/pipeline/schedule/interleaved_pp.py (hunk 3 above)
Comment (Gy-Lu) [reply colossalai/pipeline/schedule/interleaved_pp.py]: Done

File: tests/test_pipeline/test_schedule/test_interleaved.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    BATCH_SIZE = num_micro_batches
+    NUM_CHUNKS = 2
+
+    # create model
+    torch_model = MlpModel().cuda()
+
+    pp_model = copy.deepcopy(torch_model).cuda()
+
+    DP_DIM, PP_DIM, TP_DIM = 0, 1, 2
+    pg_mesh = ProcessGroupMesh(1, world_size, 1)
```
Comment (ver217) [inline tests/test_pipeline/test_schedule/test_interleaved.py:78]: Is it forced to be 3d mesh?

Comment (Gy-Lu) [reply tests/test_pipeline/test_schedule/test_interleaved.py:78]: No, in fact, I just follow the ordinary way of the training process.

**Review by ver217 — APPROVED** [review]

**Pull request merged**